# Exercise 1 (Learning to Interrupt?)


In [2]:
# to interrupt, press 'i' twice on the keyboard
while True:
    print('*', end=' ')


* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * 

KeyboardInterrupt: 

*****  
# Exercise 2 (Resetting & Restarting the Kernel)

In [3]:
print = 'PRINT'

In [1]:
# To restart kernel, press '0' twice
print('Hello World')

Hello World


******  
# Exercise 3 (Timing Stuff)

In [4]:
%%timeit
squares = [i**2 for i in range(10)] 
# print(squares)
# %timeit for i in range(10): squares

1.35 μs ± 26.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Analysis: 
- 1.35 μs is the time taken to 1 loop to finish, with an uncertainty of 26.1ns per run.
- the 26.1ns uncertainty is the standard deviation of 7,000,000 loops (7 runs x 1,000,000 loops)

In [5]:
%%timeit
squares = list()
for i in range(10):
    squares.append(i**2)
# print (squares)
# %timeit for i in range(10): squares

1.78 μs ± 135 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


Analysis:
- 1.78μs is the time taken for 1 loop to finish, with an uncertainty of 135ns per run.
- the 135ns uncertainty is the standard deviation of 7,000,000 loops (7 runs x 1,000,000 loops)

Futher thoughts:
- The 'append' in the second code causes it to have to perform an extra action, especially with more lines of code to process per cycle in the second round of code. Hence, having your code be more compact is better in general.
- However, it begs the question of whether there is the case where a single line of code processing will take longer than if it is split up into multiple lines.